### DISCUSSION TITLE
Home Assistant Button for "Scan new library files"

### DISCUSSION THREAD

**Discussion by DeltaTango69** [immich-app/immich#9888] [Q&A] [answered]
Hello,

Does anyone have an idea how to add a button in Home Assistant to scan the external library for new files?

![Screenshot 2024-05-30 104109](https://github.com/immich-app/immich/assets/82575543/8a9d084c-7734-419e-b68e-dc62484f78b9)

**Comment by DeltaTango69** [comment]
I have found this in API documenation - but with my user id it is not working.

192.168.178.5:2283/api/library/:id/scan

.:id= 
![image](https://github.com/immich-app/immich/assets/82575543/f32552cf-ce26-4ad1-b37d-d13d770e39a7)

following message I get

![image](https://github.com/immich-app/immich/assets/82575543/0fecec5d-5140-47f7-95b4-a0a527547be5)

**Reply by vincios** [reply]
That should be the library id, you can find it querying

`GET /api/library/`

**Comment by DeltaTango69** [comment]
Ok thank you there are 3 id - I have tried all but the result is the same:

How to enter in browser: 

http://192.168.178.5:2283/api/library/xxxxxxxxxxxxxxxxxxxx/scan - is that right?

[{"id":"xxxxxxxxxxxxxxxxxxxxxxxxxxxxx","ownerId":"xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx","type":"UPLOAD","name":"Default Library","createdAt":"2024-03-07T09:10:44.681Z","updatedAt":"2024-03-07T09:10:44.681Z","refreshedAt":null,"assetCount":0,"importPaths":[],"exclusionPatterns":[]},{"id":"xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx","ownerId":"xxxxxxxxxxxxxxxxxxxxxxxxxxxxx","type":"EXTERNAL","name":"New External Library","createdAt":"2024-03-07T09:12:10.611Z","updatedAt":"2024-06-01T10:00:15.153Z","refreshedAt":"2024-06-01T10:00:15.143Z","assetCount":0,"importPaths":["/mnt/media"],"exclusionPatterns":["**/*.insv","**/*.insp","**/*.dng"]}]

![image](https://github.com/immich-app/immich/assets/82575543/8ec444c0-11ac-4541-91f5-d7b5b367caf4)

**Reply by vincios** [reply, marked as answer]
This is because the [scanLibrary](https://immich.app/docs/api/scan-library) API requires a POST request

`POST /api/library/:id/scan`

so you cannot run it in your browser (that executes a GET request). You need a REST client (like [Insomnia](https://insomnia.rest/) or [Bruno](https://www.usebruno.com/), or the [RESTer](https://chromewebstore.google.com/detail/rester/eejfoncpjfgmeleakejdcanedmefagga) browser extension).

Don't forget that:
- you have to [generate](https://immich.app/docs/features/command-line-interface#obtain-the-api-key) an API Key and set it as `x-api-key` header
- The `scanLibrary` API requires two body paremeters, so you have to add them to the body of the request (as json)
  ```json
  {
    "refreshAllFiles": false,
    "refreshModifiedFiles": false
  }
  ``` 

Then, when the request works as expected, you can add it to Home Assistant using a [RESTful Command](https://www.home-assistant.io/integrations/rest_command)

**Comment by DeltaTango69** [comment]
Ok, that's one step further. I have now entered the rest command in the configuration.yaml. But I'm not sure if this is correct?

![Screenshot 2024-06-02 172754](https://github.com/immich-app/immich/assets/82575543/68d33c84-c384-4717-b92b-0e97060732a9)

Ok I have investigate a little bit more - next try seems to be also false:

![image](https://github.com/immich-app/immich/assets/82575543/cc6aefa0-4654-4400-8c7c-cdb8ac567e6c)

**Reply by Manic-87** [reply]
If I understand it correctly the `:id` is a placeholder. The `:` should be a part of it. Which means you probably just need to change the URL from `library/:289...` to `library/289...`

**Reply by DeltaTango69** [reply]
Unfortunately it also don't work. Failure from the client....

**Comment by DeltaTango69** [comment]
Ok, after changing some formating

![image](https://github.com/immich-app/immich/assets/82575543/5112cd1c-efb0-4be4-b10e-4acb42d568f0)


I got this response message:

content: ""
status: 204

What is the meaning of status 204?

**Reply by Meliox** [reply]
Indicate the server has successfully completed your request and there is no return content.

Expected reply if everything is successful if you see the response here https://immich.app/docs/api/scan-library/

**Comment by DeltaTango69** [comment]
Indeed it works now perfect :-) Now I have to take care that my switch in homeassistant, to scan the library, give some response when the return code is 204 or not. No idea in the moment to do this. But this is no question for this forum.

**Comment by DeltaTango69** [comment]
It seems to be that this api request a full scan. My wish is only to scan for new files. Maybe it is possible to revise the payload for that

```
payload = json.dumps({
  "refreshAllFiles": True,
  "refreshModifiedFiles": True
```

maybe

`"refreshNewFiles"`

Is this possible?

**Comment by jsapede** [comment]
hi all,
sorry to bring up this topic but im' struggled with making a rest command to upload a picture to immich using HA :

```
rest_command;
  immich_upload:
    url: 'http://192.168.0.145:2283/api/assets'
    method: POST
    headers:
      Accept: "application/json"
      x-api-key: "XXXXXXXXXXXXXXXX"  # replace with a valid API key
    payload: '{"assetData": "{{ assetData }}", "deviceAssetId": "{{ deviceAssetId }}", "deviceId": "{{ deviceId }}","fileCreatedAt": "{{ fileCreatedAt }}","fileModifiedAt": "{{ fileModifiedAt }}"}'
```

witch then i call from action menu with some dummy datas except for the image path: 

```
action: rest_command.immich_upload
data:
  assetData: "@/api/image/serve/6a0385c3a57798db8341f2751455964d/512x512"
  deviceAssetId: "100000875"
  deviceId: "ha"
  fileCreatedAt: "2025-01-16T14:27:52.182Z"
  fileModifiedAt: "2025-01-16T14:27:54.000Z"
```
but permanently get 400 errors : 

```
content:
  message:
    - deviceAssetId must be a string
    - deviceAssetId should not be empty
    - deviceId must be a string
    - deviceId should not be empty
    - fileCreatedAt must be a Date instance
    - fileCreatedAt should not be empty
    - fileModifiedAt must be a Date instance
    - fileModifiedAt should not be empty
  error: Bad Request
  statusCode: 400
  correlationId: lqo4e02l
status: 400
```

and in the HA logs : 

```
Error. Url: http://192.168.0.145:2283/api/assets. Status code 400. Payload: b'{\n "assetData": "@/api/image/serve/6a0385c3a57798db8341f2751455964d/512x512",\n "deviceAssetId": "1000008755",\n "deviceId": "ha",\n "fileCreatedAt": "2025-01-16T14:27:52.182Z",\n "fileModifiedAt": "2025-01-16T14:27:54.000Z"\n}'
```


event if it looks complient for me with the immich-API syntax and the python script given as example : https://immich.app/docs/guides/python-file-upload

does anyonae have an idea ?

**Reply by jsapede** [reply]
get some progress : 

```
root@ha-docker:/opt/homeassistant/config# curl -L '192.168.0.145:2283/api/assets' -H 'Content-Type: multipart/form-data' -H 'Accept: application/json' -H 'x-api-key: dBFeHRSrN92uhgem2lD9sC9Jd9AOutmrwaXVoY86g' -F 'deviceAssetId="12931249234923"' -F 'deviceId="4343452435235"' -F 'fileCreatedAt="2025-01-24T14:27:52"' -F 'fileModifiedAt="2025-01-24T14:27:52"' -F 'assetData=@"/opt/homeassistant/config/image/cbc6cf105d7a41a8274f4f9c16e79a64/original"'
{"message":"Unsupported file type original","error":"Bad Request","statusCode":400,"correlationId":"kwkpf8nd"}
```

looks like the HA converted pictures gives error but dont know why. 

same command using a jpg stored on the HA instance works perfectly using curl.

However still having trouble to pass the file path to the rest command : 

```
action: rest_command.immich_upload
data:
  assetData: "@/www/sam.jpg"
  deviceAssetId: "1000008755"
  deviceId: "ha"
  fileCreatedAt: "2025-01-23T14:27:52.182Z"
  fileModifiedAt: "2025-01-23T14:27:54.000Z"
```

gives another new error : 

```
content:
  message: File is required
  error: Bad Request
  statusCode: 400
  correlationId: 05p4rhg8
status: 400
```
